node/core/runtime-api/src/cache.rs

```diff
@@ -28,126 +27,91 @@ use polkadot_primitives::v2::{
-const VALIDATION_CODE_CACHE_SIZE: usize = 10 * 1024 * 1024;
-const CANDIDATE_PENDING_AVAILABILITY_CACHE_SIZE: usize = 64 * 1024;
-const CANDIDATE_EVENTS_CACHE_SIZE: usize = 64 * 1024;
-const SESSION_INFO_CACHE_SIZE: usize = 64 * 1024;
-const DMQ_CONTENTS_CACHE_SIZE: usize = 64 * 1024;
-const INBOUND_HRMP_CHANNELS_CACHE_SIZE: usize = 64 * 1024;
-const CURRENT_BABE_EPOCH_CACHE_SIZE: usize = 64 * 1024;
-const ON_CHAIN_VOTES_CACHE_SIZE: usize = 3 * 1024;
-const PVFS_REQUIRE_PRECHECK_SIZE: usize = 1024;
-const VALIDATION_CODE_HASH_CACHE_SIZE: usize = 64 * 1024;
-const VERSION_CACHE_SIZE: usize = 4 * 1024;
-const DISPUTES_CACHE_SIZE: usize = 64 * 1024;
-
-struct ResidentSizeOf<T>(T);
-
-impl<T: MallocSizeOf> ResidentSize for ResidentSizeOf<T> {
-	fn resident_size(&self) -> usize {
-		std::mem::size_of::<Self>() + self.0.malloc_size_of()
-	}
-}
-
-struct DoesNotAllocate<T>(T);
-
-impl<T> ResidentSize for DoesNotAllocate<T> {
-	fn resident_size(&self) -> usize {
-		std::mem::size_of::<Self>()
-	}
-}
-
-// this is an ugly workaround for `AuthorityDiscoveryId`
-// not implementing `MallocSizeOf`
-struct VecOfDoesNotAllocate<T>(Vec<T>);
-
-impl<T> ResidentSize for VecOfDoesNotAllocate<T> {
-	fn resident_size(&self) -> usize {
-		std::mem::size_of::<T>() * self.0.capacity()
-	}
-}
+const AUTHORITIES_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const VALIDATORS_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const VALIDATOR_GROUPS_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const AVAILABILITY_CORES_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const PERSISTED_VALIDATION_DATA_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const ASSUMED_VALIDATION_DATA_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const CHECK_VALIDATION_OUTPUTS_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const SESSION_INDEX_FOR_CHILD_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const VALIDATION_CODE_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const CANDIDATE_PENDING_AVAILABILITY_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const CANDIDATE_EVENTS_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const SESSION_INFO_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(128);
+const DMQ_CONTENTS_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const INBOUND_HRMP_CHANNELS_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const CURRENT_BABE_EPOCH_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const ON_CHAIN_VOTES_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const PVFS_REQUIRE_PRECHECK_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const VALIDATION_CODE_HASH_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const VERSION_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
+const DISPUTES_CACHE_CAP: Option<NonZeroUsize> = NonZeroUsize::new(1024);
```

Comment:
I wasn't sure about the cache capacities. I don't know what the expected sizes of these types are, or how the original cache sizes were chosen. I think the validation code cap should be lower, but I don't know what a safe value is. Any feedback welcome.

Reply:
As we cache per block, it would mean we cache values for 1024 different blocks. This sounds a little bit too much. 128 should be way enough or even less. We should also just have one cache size. We don't need to have a constant here per cache.